classes/CartRule.php

```diff
@@ -1553,12 +1563,47 @@ public static function cleanProductRuleIntegrity($type, $list)
     public static function getCartsRuleByCode($name, $id_lang, $extended = false)
     {
         $sql_base = 'SELECT cr.*, crl.*
-						FROM '._DB_PREFIX_.'cart_rule cr
-						LEFT JOIN '._DB_PREFIX_.'cart_rule_lang crl ON (cr.id_cart_rule = crl.id_cart_rule AND crl.id_lang = '.(int)$id_lang.')';
+						FROM ' . _DB_PREFIX_ . 'cart_rule cr
+						LEFT JOIN ' . _DB_PREFIX_ . 'cart_rule_lang crl ON (cr.id_cart_rule = crl.id_cart_rule AND crl.id_lang = ' . (int) $id_lang . ')';
         if ($extended) {
-            return Db::getInstance()->executeS('('.$sql_base.' WHERE code LIKE \'%'.pSQL($name).'%\') UNION ('.$sql_base.' WHERE name LIKE \'%'.pSQL($name).'%\')');
+            return Db::getInstance()->executeS('(' . $sql_base . ' WHERE code LIKE \'%' . pSQL($name) . '%\') UNION (' . $sql_base . ' WHERE name LIKE \'%' . pSQL($name) . '%\')');
+        } else {
+            return Db::getInstance()->executeS($sql_base . ' WHERE code LIKE \'%' . pSQL($name) . '%\'');
+        }
+    }
+
+    /**
+     * CartRules compare function to use the Product and the rules
+     *
+     * @param array $arrayProduct List Products,
+     * @param array $arrayRules List Rules,
+     * @param string $type type voucher,
+     *
+     * @return int 0 = same
+     *             1 = different
+     */
+    public function cartRulesCompare($arrayProduct, $arrayRules, $type)
```

Comment:
I think the name of this method is confusing. It filters out elements of $arrayProduct that cannot be matched to those in $arrayRules. So to me, the signature for this method should be something like this:

```php
protected function filterProducts($products, $eligibleProducts, $ruleType)
```

where
- products -> list of products from the cart
- eligibleProducts -> products that are eligible for this rule
- ruleType -> name of the rule